### PR TITLE
fix(engine): reject deployment with invalid process

### LINKED ISSUE #14028: Should not activate link catch event in subprocess
**Describe the bug**

@philippfromme discovered that a link throw event can activate a link catch event inside an (event) subprocess. See the following process.  

Screenshot from 2023-08-28 09-41-57

According to the BPMN specification and the Camunda documentation, it should not be possible to activate a link catch event that is not in the same scope as the link throw event. Both events must be in the same scope.

This issue is kind of related to https://github.com/camunda/zeebe/issues/10854.

**To Reproduce**

1. Deploy the following process 
[link-event-subprocess.bpmn](https://github.com/camunda/zeebe/files/12452163/link-event-subprocess.bpmn.txt)
2. Create a new instance of the process
3. Verify that the link catch event is activated

**Expected behavior**

The link catch event inside the (event) subprocess is not activated.

The deployment of the process is rejected because there is no link catch event in the same scope.

**Log/Stacktrace**

<details><summary>Output from Zeebe-Play</summary>
 <p>

Screenshot from 2023-08-28 10-22-54

</p>
</details>

**Environment:**
- OS: 
- Zeebe Version: `8.2.0`
- Configuration:

### PR DESCRIPTION
**Pull request by lzgabel**
## Description

The deployment of the process is rejected because there is no link catch event in the same scope.

## Related issues

closes #14028

## Definition of Done

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/ModelUtil.java [hunk 1]
```diff
@@ -138,24 +141,32 @@ public static void verifyLinkIntermediateEvents(
...
+    // link catch events names group by scope
+    final Map<BpmnModelElementInstance, Set<String>> linkCatchEventsGroupByScope =
         getEventDefinition(linkCatchEvents, LinkEventDefinition.class)
             .filter(def -> def.getName() != null && !def.getName().isEmpty())
-            .map(LinkEventDefinition::getName)
-            .collect(Collectors.toSet());
+            .collect(
+                groupingBy(
+                    LinkEventDefinition::getScope,
+                    mapping(LinkEventDefinition::getName, Collectors.toSet())));
```
Comment (nicpuppa) [inline bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/ModelUtil.java:151]: 🔧 You can extract this method, like: 

```
    private Map<BpmnModelElementInstance, Set<String>> groupEventsByScope(List<LinkEventDefinition> events) {
        return events.stream()
            .filter(def -> def.getName() != null && !def.getName().isEmpty())
            .collect(Collectors.groupingBy(
                LinkEventDefinition::getScope,
                Collectors.mapping(LinkEventDefinition::getName, Collectors.toSet())
            ));
    }
```

Then, you can use the extracted method like this:
```
Map<BpmnModelElementInstance, Set<String>> linkCatchEventsGroupByScope = groupEventsByScope(linkCatchEvents);
Map<BpmnModelElementInstance, Set<String>> linkThrowEventsGroupByScope = groupEventsByScope(linkThrowEvents);
```

By doing this we remove code duplication and improve readability 👀

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/ModelUtil.java [hunk 2]
```diff
@@ -138,24 +141,32 @@ public static void verifyLinkIntermediateEvents(
...
-    linkThrowList.forEach(
-        item -> {
-          if (!linkCatchList.contains(item)) {
-            errorCollector.accept(noPairedLinkNames(item));
+            .collect(
+                groupingBy(
+                    LinkEventDefinition::getScope,
+                    mapping(LinkEventDefinition::getName, Collectors.toSet())));
+
+    linkThrowEventsGroupByScope.forEach(
```
Comment (nicpuppa) [inline bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/ModelUtil.java:162]: 🔧 Same here, you can extract this validation in a separate private method. You can also add javadoc to explain what this validation is supposed to do.

🔧 Talking about readability, you can change this validation like this (the name of this method can be improved):

```
    private void validateLinkEvents(Map<BpmnModelElementInstance, Set<String>> throwEventsGroupByScope,
                                    Map<BpmnModelElementInstance, Set<String>> catchEventsGroupByScope,
                                    Consumer<String> errorCollector) {
        throwEventsGroupByScope.forEach((scope, throwEventNames) -> {
            Set<String> catchEventNamesForScope = catchEventsGroupByScope.getOrDefault(scope, Collections.emptySet());
            throwEventNames.stream()
                .filter(eventName -> !catchEventNamesForScope.contains(eventName))
                .forEach(eventName -> errorCollector.accept(noPairedLinkNames(eventName)));
        });
    }
```

**Review by nicpuppa — COMMENTED** [review]
Hey @lzgabel , thanks for your contribution 🚀 

I left some suggestion, please have a look 👀

**Comment by lzgabel** [comment]
Hi @nicpuppa. Please check again. I'll squash all commits when you're done checking this out. 🙇

**Comment by nicpuppa** [comment]
Hi @lzgabel , LGTM! Well done 🎉

**Review by nicpuppa — APPROVED** [review]
Hi @lzgabel , LGTM! Well done 🎉

**Comment by lzgabel** [comment]
⚡  @nicpuppa. Thanks for such a fast review! I've squashed all commits. :heart:

**Comment by nicpuppa** [comment]
bors merge

**Comment by nicpuppa** [comment]
> @nicpuppa FYI. `Link Event` is supported since version `8.2`

Thanks for raising this @lzgabel 👍

**Pull request merged**